### DISCUSSION TITLE
Compiler support syntax like Typescript JSX (inline markup)

### DISCUSSION THREAD

**Discussion by gvkhna** [withastro/roadmap#833] [Proposal]
### Body

# Summary

Support following syntax similar to React:

```jsx
const data = [<>My Fragment Text</>]
```

# Background & Motivation

- Typescript supports it so it seems an arbitrary limitation of the astro compiler, as well I don't see any reason why it couldn't or shouldn't be supported at some point. 
<img width="948" alt="Screenshot 2024-02-07 at 12 47 43 PM" src="https://github.com/withastro/roadmap/assets/222124/7cfa1087-e417-4b8f-b63a-832442d4c4b0">
- At least supporting fragments if not entire arbitrary html markup (`<p>/<span>/etc..`) would be a reasonable tradeoff.
- Useful for free-form text in astro pages, because multiline text is better formatted by prettier/etc versus strings.
- This is not a new concept, and the typescript compiler supports JSX, turning <>My Text</> to something like `new JsxFragment(JsxTextElement('My Text'))`.
- Also see https://facebook.github.io/jsx/#sec-why-not-template-literals for why template literals is not enough.

# Goals

- Support JSX syntax in astro pages and frontmatter. Ideally you could do something like the following...

```astro
---
const data = <>My Text</>
---
<Component {data} aOpt={<p>My other opt</p>} />
```

**Comment by Princesseuh** [comment]
> React supports it so it seems an arbitrary limitation of the astro compiler, as well I don't see any reason why it couldn't or shouldn't be supported at some point.

Astro and React have nothing to do with each other technically, Astro's template is not JSX despite looking like it (there's a bunch of differences)

**Reply by gvkhna** [reply]
Would be cool as hell though 👍

**Reply by fweth** [reply]
True, but there are other frameworks supporting JSX as well, like Solid, Quik, Inferno, etc. :)

But since Astro is not a JS framework, and it wants to retain compatibility with HTML, i can understand the decision. That being said, Solid's JSX is closer to HTML than React's, since they use `class` (and I guess also `for`) instead of `classList` and `htmlFor`...

**Comment by natemoo-re** [comment]
I'm going to repost some comments I shared in Discord!

> So there is actually a technical reason behind this, although the suggestion to revisit this is not out of the question…
> 
> What do you expect to see when you log that property? Currently Astro doesn’t actually have any intermediate DOM representation on the server. Slot values are not introspectable or part of our public API for that reason.
>
> The main point is that the behavior wouldn’t/couldn’t be equivalent to React. This would also get quite confusing when you’re trying to serialize props across the server/client boundary. Again, not impossible but probably not intuitive.

To be clear, `.astro` is _not_ TypeScript or JSX. It's a different language entirely that blends rules from both HTML and JSX. Astro components actually have an entirely different signature from traditional JSX components, so the `(props, ...children) => VNode` concept as well as things like `children` aren't really applicable. If you did do `console.log(Astro.props.element)` you probably wouldn't expect the result to be an `AsyncGenerator`, but that's what it would be! Our compiler mostly generates code to do string concatenation because `.astro` components only need to run on the server. It's less flexible than a full VDOM representation, but it is more efficient.

This proposal would be a massive undertaking that exposes an entirely new API surface for us to worry about. It's still entirely valid! It's just a question of whether the benefits justify the engineering effort required.

**Reply by gvkhna** [reply]
Maybe just text fragments then?

**Reply by mb21** [reply]
related: [Support JSX syntax for Astro components](https://github.com/withastro/roadmap/discussions/716)

Could investigate an existing streaming JSX parser like Solid's?

**Comment by gvkhna** [comment]
Changed rfc to typescript, no mention of react, typescript has this well defined in in it's lib. That was my initial catalyst for wondering why this doesn't actually work because I could import the JsxFragment type etc.

**Reply by Princesseuh** [reply]
I suspect that you might be misunderstanding how everything work together, TypeScript having support for JSX has nothing to do with how Astro works, as not only is Astro neither TypeScript, or JSX, but it's also purely a concept specific to TypeScript's compiler. 

It's not a type you can simply make and pass around, it's either TypeScript's internal representation of a JSX node (which has nothing to do with Astro, since Astro is not JSX!) or a compiled form specific to a specific JSX framework

**Reply by gvkhna** [reply]
I understand, I just think it would be a great addition but I guess it's technically infeasible with Astros arch. A shame but not a show stopper so I understand low priority as well.
